README.md
<p align="center">
  <img width="240" src="https://raw.githubusercontent.com/exoplanet-dev/exoplanet/main/docs/_static/logo.png">
  <br><br>
  <a href="https://github.com/exoplanet-dev/exoplanet/actions/workflows/tests.yml">
    <img src="https://github.com/exoplanet-dev/exoplanet/actions/workflows/tests.yml/badge.svg" alt="Tests">
  </a>
  <a href="https://docs.exoplanet.codes">
    <img src="https://readthedocs.org/projects/exoplanet/badge/?version=latest" alt="Docs">
  </a>
  <a href="https://coveralls.io/github/exoplanet-dev/exoplanet?branch=main">
    <img src="https://coveralls.io/repos/github/exoplanet-dev/exoplanet/badge.svg?branch=main" alt="Coverage">
  </a>
</p>

# exoplanet

Fast & scalable MCMC for all your exoplanet needs! _exoplanet_ is a toolkit for
probabilistic modeling of time series data in astronomy with a focus on
observations of [exoplanets](https://en.wikipedia.org/wiki/Exoplanet), using
[PyMC3](https://docs.pymc.io). _PyMC3_ is a flexible and high-performance model
building language and inference engine that scales well to problems with a large
number of parameters. _exoplanet_ extends _PyMC3_'s language to support many of
the custom functions and distributions required when fitting exoplanet datasets.

Read the full documentation at [docs.exoplanet.codes](https://docs.exoplanet.codes).

## Installation

The quickest way to get started is to use [pip](https://pip.pypa.io):

```bash
python -m pip install exoplanet
```

Note that you will need Python (>=3.6) installed for this to work, but then this
will install all the required dependencies.

Check out the [main installation documentation
page](https://docs.exoplanet.codes/en/latest/user/install/) for more options.

## Usage

Check out the tutorials and API docs on [the docs
page](https://docs.exoplanet.codes) for example usage and much more info. You
can also find more in-depth examples on the [exoplanet case studies
page](https://gallery.exoplanet.codes).

## Contributing

_exoplanet_ is an open source project and we would love it if you wanted to
contribute. Check out [the developer
documentation](https://docs.exoplanet.codes/en/latest/user/dev/) for more info
about getting started.


HISTORY.rst
0.5.1 (2021-06-21)
++++++++++++++++++

- Adds support for testing using tox (`#197 <https://github.com/exoplanet-dev/exoplanet/pull/197>`_)
- Updates documentation, tutorials, and case studies for JOSS review (`#186 <https://github.com/exoplanet-dev/exoplanet/pull/186>`_ and `#200 <https://github.com/exoplanet-dev/exoplanet/pull/200>`_)


0.5.0 (2021-05-04)
++++++++++++++++++

- Adds high-level, pure-Theano/Aesara implementation of ``RegularGridInterpolator`` to replace compiled version (`#167 <https://github.com/exoplanet-dev/exoplanet/pull/167>`_)
- Moves compiled Ops to `exoplanet-core <https://github.com/exoplanet-dev/exoplanet-core>`_ making exoplanet a pure-Python package (`#171 <https://github.com/exoplanet-dev/exoplanet/pull/171>`_)
- Moves `ReboundOp` and `ReboundOrbit` to `rebound-pymc3 <https://github.com/exoplanet-dev/rebound-pymc3>`_ (`#166 <https://github.com/exoplanet-dev/exoplanet/pull/166>`_)
- Removes deprecated submodules (``gp``, ``optimize``, and ``sampling``) and base distributions (e.g. ``xo.Angle``) (`#165 <https://github.com/exoplanet-dev/exoplanet/pull/165>`_)


0.4.5 (2021-03-15)
++++++++++++++++++

- Adds support for recent versions of PyMC3 built on Aesara (`#150 <https://github.com/exoplanet-dev/exoplanet/pull/150>`_)
- Fixes bug in `SimpleTransitOrbit` contact point calculation (`#148 <https://github.com/exoplanet-dev/exoplanet/pull/148>`_)


0.4.4 (2021-01-04)
++++++++++++++++++

- Adds support for relativistic orbits using REBOUNDx (`#98 <https://github.com/exoplanet-dev/exoplanet/pull/98>`_)


0.4.3 (2020-12-08)
++++++++++++++++++

- Adds support for PyMC3 v3.10 and Theano-pymc


0.4.2 (2020-11-15)
++++++++++++++++++

- Fixes issue with automatic execution of tutorials


0.4.1 (2020-11-15)
++++++++++++++++++

- Fixes pickling error when sampling in parallel (`#120 <https://github.com/exoplanet-dev/exoplanet/pull/120>`_)


0.4.0 (2020-10-05)
++++++++++++++++++

- Adds faster solver for Kepler's equation
- Improves automation of documentation and release process
- Removes most C++ Ops and replaces them with pre-compiled functions which allows the distribution of binary wheels
- Removes ``gp`` submodule; moved to new `celerite2 <https://celerite2.readthedocs.io>`_ package
- Deprecates some non-exoplanet-specific functions and distributions; moved to `pymc3-ext <https://github.com/exoplanet-dev/pymc3-ext>`_ package


0.3.3 (2020-09-09)
++++++++++++++++++

- Fixes compatibility with PyMC3 version 3.9


0.3.2 (2020-05-04)
++++++++++++++++++

- Fixes a documentation bug introduced in v0.3.1


0.3.1 (2020-05-04)
++++++++++++++++++

- Adds support for light travel time when computing positions in Keplerian orbits
- Adds ``SecondaryEclipseLightCurve`` for modeling eclipsing binaries
- Adds ``UnitDisk`` distribution for fitting eccentricity vectors


0.3.0 (2020-04-03)
++++++++++++++++++

- Adds tests and support for Windows
- Adds a "Jacobian" interface to the orbit objects for reparameterization


0.2.6 (2020-03-23)
++++++++++++++++++

- Adds support for fitting circular orbits with duration
- Adds a ``bls_estimator`` for transit search using Astropy's ``BoxLeastSquares``


0.2.5 (2020-03-11)
++++++++++++++++++

- Improves infrastructure for generating documentation
- Adds an ``EclipsingBinaryLightCurve`` for building binary star models
- Adds ``DensityDist`` implementation for celerite GP likelihoods


0.2.4 (2019-12-30)
++++++++++++++++++

- Makes ``rebound`` an optional dependency


0.2.3 (2019-11-12)
++++++++++++++++++

- Adds ``ConditionalMeanOp`` and ``DotLOp`` for scalable conditional mean calculation
  and prior sampling with celerite
- Adds developer documentation
- Moves documentation to a separate repository


0.2.2 (2019-10-25)
++++++++++++++++++

- Adds ``TTVOrbit`` tutorial
- Switches tutorials to `lightkurve <https://docs.lightkurve.org>`_ for data access
- Improves packaging and code style features
- Fixes bugs and improves interface to ``TTVOrbit``


0.2.1 (2019-09-26)
++++++++++++++++++

- Adds a new interface for tuning dense mass matrices with less overhead
- Adds support for photodynamics using `rebound <https://rebound.rtfd.io>`_
- Adds a new interface for assigning units to Theano variables
- Adds new physically-motivated distributions for impact parameter and
  eccentricity
- Improves test coverage
- Fixes bug in diagonal elements of the ``IntegratedTerm`` model
- Fixes bug in indexing for ``TTVOrbit`` models with large TTVs


0.2.0 (2019-08-04)
++++++++++++++++++

- Updates ``starry`` to get much better performance for high order spherical
  harmonics
- Renames ``StarryLightCurve`` to ``LimbDarkLightCurve``
- Restructures the C++ backend to reduce code duplication
- Adds support for fitting of astrometric observations
- Adds support for exposure time integration in ``celerite`` models
- Adds new distributions for periodic parameters and U(0, 1).
- Fixes many small bugs


0.1.6 (2019-04-24)
++++++++++++++++++

- Fixes some edge case failures for Kepler solver
- Improves reliability of contact point solver and fails (more) gracefully
  when this doesn't work; this reduces the number of divergences when fitting
  a transit model


0.1.5 (2019-03-07)
++++++++++++++++++

- Improves contact point solver using companion matrix to solve quadratic
- Improves reliability of ``Angle`` distribution when the value of the angle
  is well constrained


0.1.4 (2019-02-10)
++++++++++++++++++

- Improves the reliability of the PyMC3Sampler
- Adds a new ``optimize`` function since the ``find_MAP`` method
  in PyMC3 is deprecated
- Adds cronjob script for automatically updating tutorials.


0.1.3 (2019-01-09)
++++++++++++++++++

- Adds a more robust and faster Kepler solver (`ref
  <http://adsabs.harvard.edu/abs/1991CeMDA..51..319N>`_)
- Fixes minor behavioral bugs in PyMC3 sampler wrapper


0.1.2 (2018-12-13)
++++++++++++++++++

- Adds regular grid interpolation Op for Theano
- Fixes major bug in handling of the stellar radius for transits

multi.py

# coding: utf-8

import kplr
import numpy as np
import matplotlib.pyplot as plt
# from astropy.table import Table
from scipy.signal import medfilt

import tensorflow as tf
session = tf.InteractiveSession()

from exoplanet import transit, kepler, interp

import hemcee
import tfmodeling as tfm

# import celeriteflow as cf


# In[2]:

kepid = 11709124
client = kplr.API()
star = client.star(kepid)
t = []
y = []
yerr = []
quarter = []
texp = []
for i, lc in enumerate(star.get_light_curves(short_cadence=False)):
    with lc.open() as f:
        hdr = f[1].header
        texp0 = hdr["INT_TIME"] * hdr["NUM_FRM"] / (24. * 60. * 60.)
        data = f[1].data

    q = data["SAP_QUALITY"]
    t0 = data["TIME"]
    y0 = data["PDCSAP_FLUX"]
    m = (q == 0) & np.isfinite(t0) & np.isfinite(y0)
    mu = medfilt(y0[m], 101)
#     mu = np.median(y0[m])
    t.append(t0[m])
    y.append((y0[m] / mu - 1.0) * 1e3)
    yerr.append(data["PDCSAP_FLUX_ERR"][m] * 1e3 / mu)
    quarter.append(i + np.zeros(m.sum(), dtype=int))
    texp.append(texp0 + np.zeros(m.sum()))
t = np.ascontiguousarray(np.concatenate(t), dtype=np.float64)
y = np.ascontiguousarray(np.concatenate(y), dtype=np.float64)
yerr = np.ascontiguousarray(np.concatenate(yerr), dtype=np.float64)
quarter = np.ascontiguousarray(np.concatenate(quarter), dtype=int)
texp = np.median(np.concatenate(texp))


# In[3]:

kois = star.kois
spec = np.empty(len(kois),
                dtype=[("period", float), ("ror", float), ("t0", float), ("duration", float)])
for i, koi in enumerate(kois):
    spec["period"][i] = koi.koi_period
    spec["ror"][i] = koi.koi_ror
    spec["t0"][i] = koi.koi_time0bk % koi.koi_period
    spec["duration"][i] = koi.koi_duration / 24.0
spec = spec[np.argsort(spec["period"])]


# In[20]:

# gkf = Table.read("kepler_dr2_1arcsec.fits", format="fits").to_pandas()
# star = gkf[gkf.kepid.astype(int) == kepid].iloc[0]

# vals = dict()
# for k in ["radius", "mass"]:
#     mn = star[k] + star[k + "_err2"]
#     mx = star[k] + star[k + "_err1"]
#     log_mu = np.log(star[k])
#     log_err = 0.5 * ((log_mu - np.log(mn)) + (np.log(mx) - log_mu))
#     vals[k] = (log_mu, log_err)

# np.random.seed(42)
# N = 50000
# R = np.exp(vals["radius"][0] + vals["radius"][1] * np.random.randn(N))
# M = np.exp(vals["mass"][0] + vals["mass"][1] * np.random.randn(N))
# log_dens = np.log(0.75 * M / (np.pi * R**3))
# log_dens = (np.mean(log_dens), np.std(log_dens))
# print(log_dens)
log_dens = -1.4547524087769572, 0.3502230916468015


# In[21]:

t0 = spec["t0"][-1]
tau = spec["duration"][-1]
rng = (t0 - 1000 < t) & (t < t0 + 1000)
plt.plot(t[rng], y[rng], "k")


# In[6]:

spec


# In[7]:

T = tf.float64
np.random.seed(42)

G_grav = 2945.462538537765
factor = (G_grav / (3 * np.pi)) ** (1./3)

t_tensor = tf.constant(t[rng], dtype=T)
y_tensor = tf.constant(y[rng], dtype=T)
yerr_tensor = tf.constant(yerr[rng], dtype=T)

# Star
log_jitter = tfm.Parameter(
    2.0*np.log(np.median(yerr)),
    bounds=(2.0*np.log(0.01*np.median(yerr)), 2.0*np.log(10.0*np.median(yerr))),
    dtype=T)
flux = tfm.Parameter(0.0, bounds=(-2, 2), dtype=T)
q = tfm.Parameter(0.5 + np.zeros(2), bounds=(0, 1), dtype=T)
log_rho_star = tfm.Parameter(log_dens[0], dtype=T)
q1 = q.value[0]
q2 = q.value[1]
c1 = 2.0 * tf.sqrt(q1) * q2
c2 = tf.sqrt(q1) * (1.0 - 2.0 * q2)
ld = transit.QuadraticLimbDarkening(c1, c2)

# planets
pls = [2, 5]
n_pl = len(pls)
log_Ps = []
for p in spec["period"][pls]:
#     bounds = None
    bounds = (np.log(p - 1), np.log(p + 1))
    if p > 365:
        bounds = (np.log(10), np.log(100*365.))
    log_Ps.append(tfm.Parameter(np.log(p), bounds=bounds, dtype=T))
log_P = tf.stack([p.value for p in log_Ps])

log_r = tfm.Parameter(np.log(spec["ror"][pls]), dtype=T)  # bounds=(-10, 10), dtype=T)
t0 = tfm.Parameter(spec["t0"][pls], dtype=T)
b = tfm.Parameter(0.5 + np.zeros(n_pl), bounds=(-2.0, 2.0), dtype=T)
e = tfm.Parameter(np.random.uniform(0, 0.01, n_pl), bounds=(0, 1.0), dtype=T)
# e = tfm.Parameter(np.zeros(n_pl), dtype=T)
omega_vec = tfm.UnitVector(np.random.randn(n_pl, 2), dtype=T)

period = tf.expand_dims(tf.exp(log_P), 1)
ror = tf.exp(log_r.value)
a = factor * tf.exp((2.0 * log_P + log_rho_star.value) / 3.0)
cosi = tf.abs(b.value) / a
incl = tf.acos(cosi)

sin_omega = omega_vec.value[:, 0]
cos_omega = omega_vec.value[:, 1]
omega = tf.atan2(sin_omega, cos_omega)

# Set up the interpolation grid for each planet
nint = 7
dt = texp * np.linspace(-0.5, 0.5, 2*nint+1)[1:-1:2]
hp = 0.5 * period
grid = 1.5 * spec["duration"][pls][:, None] * np.linspace(-0.5, 0.5, int(np.ceil(5 * spec["duration"].max() / texp)))[None, :]
grid_phase = tf.constant(grid, dtype=T) + hp
grid = tf.constant(grid[:, :, None] + dt[None, None, :], dtype=T)

# Solve Kepler's equation on the grid
coords = a[None, :, None, None] * kepler.sky_position(
    period[:, :, None],
    tf.zeros((n_pl, 1, 1), dtype=T),  # <- t0
    e.value[:, None, None],
    omega[:, None, None],
    incl[:, None, None],
    grid)

# Compute the sky position and the transit light curve on the grid
d = tf.sqrt(tf.reduce_sum(tf.square(coords[:2]), axis=0))
delta = transit.transit_depth(ld, d, ror[:, None, None]+tf.zeros_like(d), direction=coords[2], n_integrate=100)
transit_grid = 1e3 * tf.reduce_mean(delta, axis=-1)

# Compute the time coordinate for each planet
arg = tf.add(t_tensor[None, :], (-t0.value[:, None] + hp))
phase = tf.mod(arg, period)   # - hp[:, None]
# p = period[:, None]  # -tf.ones(n_pl, dtype=T)
transit_model = flux.value - tf.reduce_sum(interp.interp1d(phase, grid_phase, transit_grid), axis=0)

# transit_model = flux.value - 1e3 * tf.reduce_mean(tf.reduce_sum(delta, axis=0), axis=1)

err2 = tf.square(yerr_tensor) + tf.exp(log_jitter.value)
chi2 = tf.reduce_sum(tf.square(transit_model - y_tensor) / err2 + tf.log(err2))
chi2 += tf.square((log_rho_star.value - log_dens[0]) / log_dens[1])

# Beta prior
alpha = 0.867
beta = 3.03
chi2 -= 2*tf.reduce_sum((alpha-1)*tf.log(e.value) + (beta-1)*tf.log(1-e.value))

log_prob = -0.5*chi2

params = [flux, log_jitter, q, log_rho_star] + log_Ps + [log_r, t0, b, e, omega_vec]
model = tfm.Model(log_prob, params, session=session)
grad = tf.gradients(log_prob, model.get_parameters())

opt = tf.contrib.opt.ScipyOptimizerInterface(chi2, model.get_parameters())

session.run(tf.global_variables_initializer())

print(session.run(grad))


# In[8]:

arg.shape, period.shape, phase.shape


# In[19]:

from tensorflow.python.client import timeline

session.run(grad)

options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
run_metadata = tf.RunMetadata()

session.run(grad, options=options, run_metadata=run_metadata)
session.run(grad, options=options, run_metadata=run_metadata)

fetched_timeline = timeline.Timeline(run_metadata.step_stats)
chrome_trace = fetched_timeline.generate_chrome_trace_format()
with open('timeline_01.json', 'w') as f:
    f.write(chrome_trace)

get_ipython().magic('timeit session.run(grad)')
